interfaces/builtin/pwm.go

```diff
@@ -0,0 +1,144 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/systemd"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+)
+
+// https://www.kernel.org/doc/Documentation/pwm.txt
+const pwmSummary = `allows access to specifc PWM channel`
+
+const pwmBaseDeclarationSlots = `
+  pwm:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
+    deny-auto-connection: true
+`
+
+var pwmSysfsPwmChipBase = "/sys/class/pwm/pwmchip%d"
+
+// pwmInterface type
+type pwmInterface struct {
+	commonInterface
+}
+
+// BeforePrepareSlot checks the slot definition is valid
+func (iface *pwmInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	// must have a PWM chanel number
+	number, ok := slot.Attrs["number"]
```

Comment:
is this number the:

per-chip relative index of the PWM device

or the 

global index of the PWM device

(I'm looking at https://www.kernel.org/doc/html/latest/driver-api/miscellaneous.html#c.pwm_device)

Reply:
@pedronis It's per-chip index to channel exactly. For example, I take the test code to use:

```
slots:
    my-pin:
        interface: pwm
        chip-number: 10
        number: 100
```
so that required index (aka number) of chip10 under sysfs will be `/sys/class/pwm/pwmchip10/pwm100/`.

Reply:
@pedronis It's per-chip index to channel exactly. For example, I take the test code to use:

```
slots:
    my-pin:
        interface: pwm
        chip-number: 10
        number: 100
```
so that required index (aka number here) of chip10 under sysfs will be `/sys/class/pwm/pwmchip10/pwm100/`.

From above description, `pwm_chip` provides linked-list to point `pwm_device` which exposed one and more PWM signal sources, so if I understand correctly the relationship between PWM interface and document should be:

per-chip relative index of the PWM device -> number
global index of the PWM device -> chip-number

Reply:
Maybe given an idea roughly:

number -> perchip-index
chip-number -> global-index

Reply:
what about an attribute of `channel` instead? so then we have:
```
slots:
  my-pin:
    interface: pwm
    chip-number: 10
    channel: 100
```
?

Reply:
I'm ok with above +1, @pedronis, can we finalize that?

Reply:
channel is fine with me (it has a meaning for snaps but given the context is should be fine)

Reply:
Done for change from `number` to `channel`. Thx.